consensus/core/src/base_committer.rs

```diff
@@ -0,0 +1,381 @@
+        leaders: impl Iterator<Item = &'a LeaderStatus>,
+    ) -> LeaderStatus {
+        // The anchor is the first committed leader with round higher than the decision round of the
+        // target leader. We must stop the iteration upon encountering an undecided leader.
+        let anchors = leaders.filter(|x| leader_slot.round + self.options.wave_length <= x.round());
+
+        for anchor in anchors {
+            tracing::trace!(
+                "[{self}] Trying to indirect-decide {leader_slot} using anchor {anchor}",
+            );
+            match anchor {
+                LeaderStatus::Commit(anchor) => {
+                    return self.decide_leader_from_anchor(anchor, leader_slot);
+                }
+                LeaderStatus::Skip(..) => (),
+                LeaderStatus::Undecided(..) => break,
+            }
+        }
+
+        LeaderStatus::Undecided(leader_slot)
+    }
+
+    pub fn elect_leader(&self, round: Round) -> Option<Slot> {
+        let wave = self.wave_number(round);
+        tracing::debug!(
+            "elect_leader: round={}, wave={}, leader_round={}, leader_offset={}",
+            round,
+            wave,
+            self.leader_round(wave),
+            self.options.leader_offset
+        );
+        if self.leader_round(wave) != round {
+            return None;
+        }
+
+        Some(Slot::new(
+            self.leader_schedule
+                .elect_leader(round, self.options.leader_offset),
+            round,
+        ))
+    }
+
+    /// Return the wave in which the specified round belongs. This takes into
+    /// account the round offset for when pipelining is enabled.
+    fn wave_number(&self, round: Round) -> WaveNumber {
+        round.saturating_sub(self.options.round_offset) / self.options.wave_length
+    }
+
+    /// Return the leader round of the specified wave. The leader round is always
+    /// the first round of the wave. This takes into account round offset for when
+    /// pipelining is enabled.
+    fn leader_round(&self, wave: WaveNumber) -> Round {
+        (wave * self.options.wave_length) + self.options.round_offset
+    }
+
+    /// Return the decision round of the specified wave. The decision round is
+    /// always the last round of the wave. This takes into account round offset
+    /// for when pipelining is enabled.
+    fn decision_round(&self, wave: WaveNumber) -> Round {
+        let wave_length = self.options.wave_length;
+        (wave * wave_length) + wave_length - 1 + self.options.round_offset
+    }
+
+    /// Find which block is supported at (author, round) by the given block.
+    /// Blocks can indirectly reference multiple other blocks at (author, round),
+    /// but only one block at (author, round) will be supported by the given block.
+    /// If block A supports B at (author, round), it is guaranteed that any
+    /// processed block by the same author that directly or indirectly includes
+    /// A will also support B at (author, round).
```

Comment:
Now that we use the `Slot` probably this can be replaced on the comments as well wherever we refer to `(author, round)` ?